include/oneapi/tbb/detail/_task.h

```diff
@@ -103,7 +103,7 @@ class wait_context {
 
     void add_reference(std::int64_t delta) {
         call_itt_task_notify(releasing, this);
-        std::uint64_t r = m_ref_count.fetch_add(delta) + delta;
+        std::uint64_t r = m_ref_count.fetch_add(static_cast<std::uint64_t>(delta)) + static_cast<std::uint64_t>(delta);
```

Comment:
It is unsafe to do so because delta can be negative. https://github.com/oneapi-src/oneTBB/blob/3b161438c9bb4099b41838fda21dbfcb02980662/include/oneapi/tbb/detail/_task.h#L146

Reply:
I suspected that is a problem, but according to the compiler, the code is already unsafe.  If I simply do:

```
        std::uint64_t r = static_cast<std::uint64_t>(m_ref_count.fetch_add(delta) + delta);
```
the original warnings persist:

```
Showing All Messages
/tbb/include/oneapi/tbb/detail/_task.h:106:85: Implicit conversion changes signedness: 'std::int64_t' (aka 'long long') to 'unsigned long long'

/tbb/include/oneapi/tbb/detail/_task.h:106:76: Implicit conversion changes signedness: 'std::int64_t' (aka 'long long') to 'unsigned long long'
```

I believe that this is due to the definition of `m_ref_count`:
```
    std::atomic<std::uint64_t> m_ref_count{};
```
If I change `m_ref_count` to:
```
    std::atomic<std::int64_t> m_ref_count{};
```
then the warnings above are suppressed, but a new warning is triggered in `continue_execution` for the opposite reason.

Obviously, I'm not familiar with this code, but the following suppresses all warnings:
```
    std::atomic<std::uint64_t> m_ref_count{};
    std::atomic<std::int64_t> m_ref_count_signed{};

    void add_reference(std::int64_t delta) {
        call_itt_task_notify(releasing, this);
        std::uint64_t r = static_cast<std::uint64_t>(m_ref_count_signed.fetch_add(delta) + delta);

        __TBB_ASSERT_EX((r & overflow_mask) == 0, "Overflow is detected");

        if (!r) {
            // Some external waiters or coroutine waiters sleep in wait list
            // Should to notify them that work is done
            std::uintptr_t wait_ctx_addr = std::uintptr_t(this);
            r1::notify_waiters(wait_ctx_addr);
        }
    }

    bool continue_execution() const {
        std::uint64_t r = m_ref_count.load(std::memory_order_acquire);
        __TBB_ASSERT_EX((r & overflow_mask) == 0, "Overflow is detected");
        return r > 0;
    }
```

Would this change work?

Reply:
You right, that there is already a implicit conversion, thank you for pointing that. So everything should be ok with explicit cast.

Reply:
Thank you for clarifying.  I don't understand the code, but hopefully the casts are at least explicit now.